rad/prelude/cmd-parsing.rad

```diff
@@ -27,32 +30,60 @@
+    (def-rec parse-cli
+      (fn [args]
+        (def parse-arg
+          (fn [opt args]
+            (def opt-key (lookup :key opt))
+            (match args
+              (/cons 'val 'rest)
+                  (if (elem? val (lookup :possible-values opt))
+                    (modify-map opt-key (fn [x] (cons val x)) (parse-cli rest))
+                    (parse-failure (string-append "Invalid argument " val " for option " (show opt-key) ".") help))
+              []  (parse-failure (string-append "Missing argument for option " (show opt-key) ".") help)
+              )))
+        (def parse-opt
+          (fn [opt rest]
+            (def option
+              (match (filter (fn [x] (elem? opt (lookup :options x))) cmd-opts)
+                []    (parse-failure (string-append "Invalid option " opt " for command list.") help)
+                'filt (first filt)))
+            (if (eq? (lookup :type option) :flag)
+              (insert (lookup :key option) #t (parse-cli rest))
+              (parse-arg option rest))))
+        (match args
+          /nil                default
+          (/cons 'opt 'rest)  (parse-opt opt rest))))
     (fn [v]
       (match v
-        [cmd]           [:just {opts []}]
-        (/cons cmd 'os) [:just {opts (opts-map os)}]
+        [cmd]           [:just {opts default}]
+        (/cons cmd 'os) [:just {opts (parse-cli os)}]
         _               :nothing))))
 
-(:test "/cmd-1-opts"
+(:test "/cmd-opts"
   [:setup
     (do
-      (def opts-map
-        (fn [opts]
-          (map
-            (fn [opt]
-              (if (elem? opt ["--new" "--old"])
-                (drop 2 opt)
-                "unknown filter")) opts))))]
-  [ ((/cmd-1-opts "list" 'opts opts-map) ["list"]) ==> [:just {opts []}]]
-  [ ((/cmd-1-opts "list" 'opts opts-map) ["no-list"]) ==> :nothing]
-  [ ((/cmd-1-opts "list" 'opts opts-map) ["list" "--old"]) ==> [:just {opts ["old"]}]]
-  [ ((/cmd-1-opts "list" 'opts opts-map) ["list" "--old" "--new"]) ==> [:just {opts ["old" "new"]}]]
-  [ ((/cmd-1-opts "list" 'opts opts-map) ["list" "--filter"]) ==> [:just {opts ["unknown filter"]}]]
-  [ ((/cmd-1-opts "list" 'opts opts-map) ["list" "--old" "--current"]) ==> [:just {opts ["old" "unknown filter"]}]])
+      (def help "help")
+      (def cmd-opts
+        [
+          { :key :state :type :multi-opt :options ["-s" "--state" "--filter-by-state"] :possible-values ["acc" "prop"] :default []}
+          { :key :fancy :type :flag :options ["--fancy"] :default #f }
+        ])
+      (def /cmd (/cmd-opts "list" 'opts cmd-opts help)))]
+  [(/cmd ["list"] ) ==> [:just {opts {:state [] :fancy #f}}]]
+  [(/cmd ["list" "--fancy"] ) ==> [:just {opts {:state [] :fancy #t}}]]
+  [(/cmd ["list" "--state" "acc"] ) ==> [:just {opts {:state ["acc"] :fancy #f}}]]
+  [(/cmd ["list" "--fancy" "-s" "acc"] ) ==> [:just {opts {:state ["acc"] :fancy #t}}]]
+  [(/cmd ["list" "--state" "prop" "-s" "acc"] ) ==> [:just {opts {:state ["prop" "acc"] :fancy #f}}]])
```

Comment:
Could you add test cases for the parse failures? We should test an invalid option, a missing argument, an invalid option argument, and a positional argument (i.e `(/cmd ["list" "arg"])`.